### PR TITLE
Chronos: support `TSDataset` input and add `from_tsdataset` method for `Forecaster`

### PR DESCRIPTION
**Pull request by liangs6212**
## Description

This PR add

1. a new way (`from_tsdataset`) to initialize a Forecaster (including pytorch-based TCN, LSTM, S2S, NBeats).
2. support `TSDataset` as an input in `fit`, `evaluate`, `predict`, `evaluate_with_onnx`, `predict_with_onnx`.

similar API in tf based forecaster and autoformer will be done in other PRs.

## Motivation and Context

Many users feel confused when they initialize a new forecaster. They can't understand how to set `input_feature_num`, `output_feature_num`, `past_seq_len` and `future_seq_len`. These parameters can be set automatically by input data.

Some users suggested that using a `TSDataset` as input data might be more stright-forward.

inspired by: `from_dataset` in `pytorch-forecasting`: https://pytorch-forecasting.readthedocs.io/en/stable/api/pytorch_forecasting.models.nbeats.NBeats.html#pytorch_forecasting.models.nbeats.NBeats.from_dataset

## API Usage or Code Design.

Here is a very typical usage after this PR.

```python
from bigdl.chronos.forecaster import TCNForecaster
from bigdl.chronos.data import TSDataset

if __name__ == "__main__":
    tsdata = TSDataset.from_pandas(df, target_col=["value"], dt_col="date")
    tsdata.roll(lookback=10, horizon=1)
    loader = tsdata.to_torch_data_loader()
    forecaster = TCNForecaster.from_tsdataset(tsdata)
    forecaster.fit(tsdata/loader)
    forecaster.evaluate(tsdata/loader)
```

Second possible usage:

```python
from bigdl.chronos.forecaster import TCNForecaster
from bigdl.chronos.data import TSDataset

if __name__ == "__main__":
    tsdata = TSDataset.from_pandas(df, target_col=["value"], dt_col="date")
    loader = tsdata.to_torch_data_loader(lookback=10, horizon=1, roll=True)
    forecaster = TCNForecaster.from_tsdataset(tsdata)
    forecaster.fit(loader)  # use tsdata here is also valid
    forecaster.evaluate(loader)  # use tsdata here is also valid
```
Third possible usage:

```python
from bigdl.chronos.forecaster import TCNForecaster
from bigdl.chronos.data import TSDataset

if __name__ == '__main__':
    tsdata = TSDataset.from_pandas(df, target_col=["value"], dt_col="date")
    forecaster = TCNForecaster.from_tsdataset(tsdata, past_seq_len=24, future_seq_len=5)
    forecaster.fit(tsdata)  # use tsdata here is also valid
    forecaster.evaluate(tsdata)  # use tsdata here is also valid
```

## Related Link or Issue

https://github.com/intel-analytics/BigDL/issues/4772

## How was this PR tested?
unit tests is added

- [ ] jenkins passed (please provide link):

## License & Dependency
N/A

### CODE REVIEW

File: python/chronos/src/bigdl/chronos/forecaster/lstm_forecaster.py [hunk 1]
```diff
@@ -23,11 +23,16 @@ class LSTMForecaster(BasePytorchForecaster):
     """
     Example:
         >>> #The dataset is split into x_train, x_val, x_test, y_train, y_val, y_test
+        >>> # 1. Initialize Forecaster directly
         >>> forecaster = LSTMForecaster(past_seq_len=24,
                                         input_feature_num=2,
                                         output_feature_num=2,
                                         ...)
-        >>> forecaster.fit((x_train, y_train))
+        >>> # 2. The from_dataset method can also initialize a LSTMForecaster.
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/forecaster/lstm_forecaster.py:31]: from_tsdataset

File: python/chronos/src/bigdl/chronos/forecaster/lstm_forecaster.py [hunk 2]
```diff
@@ -140,3 +145,33 @@ def customized_loss_creator(config):
         self.use_hpo = False
 
         super().__init__()
+
+    @staticmethod
+    def from_tsdataset(tsdataset, past_seq_len=None, **kwargs):
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/forecaster/lstm_forecaster.py:150]: try put this method to base_forecaster and make it a classmethod

File: python/chronos/src/bigdl/chronos/forecaster/base_forecaster.py [hunk 3]
```diff
@@ -914,6 +960,40 @@ def quantize(self, calib_data=None,
...
+            past_seq_len = tsdataset.numpy_x.shape[1]
+            future_seq_len = tsdataset.numpy_y.shape[1]
+        # TODO Support specify 'auto' as past_seq_len.
+        if all([past_seq_len is None, future_seq_len is None, tsdataset.numpy_x is None]):
+            from bigdl.nano.utils.log4Error import invalidInputError
+            invalidInputError(False,
+                              "Forecaster requires 'past_seq_len' and 'future_seq_len' to specify "
+                              "the history time step and output time step.")
+        output_feature_num = len(tsdataset.target_col)
+        input_feature_num = output_feature_num + len(tsdataset.feature_col)
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/forecaster/base_forecaster.py:990]: self.roll_target, self.roll_feature

File: python/chronos/src/bigdl/chronos/forecaster/lstm_forecaster.py [hunk 4]
```diff
@@ -23,13 +23,18 @@ class LSTMForecaster(BasePytorchForecaster):
     """
     Example:
         >>> #The dataset is split into x_train, x_val, x_test, y_train, y_val, y_test
+        >>> # 1. Initialize Forecaster directly
         >>> forecaster = LSTMForecaster(past_seq_len=24,
                                         input_feature_num=2,
                                         output_feature_num=2,
                                         ...)
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/forecaster/lstm_forecaster.py:30]: add a blank line

File: python/chronos/src/bigdl/chronos/forecaster/lstm_forecaster.py [hunk 5]
```diff
@@ -23,13 +23,18 @@ class LSTMForecaster(BasePytorchForecaster):
...
-        >>> forecaster.fit((x_train, y_train))
+        >>> # 2. Initialize Forecaster from from_tsdataset
+        >>> tsdata = TSDataset.from_pandas(df, ...)
+        >>> tsdata.roll(lookback=24, horizon=1)
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/forecaster/lstm_forecaster.py:33]: line 32-33 can be deleted

File: python/chronos/src/bigdl/chronos/forecaster/lstm_forecaster.py [hunk 6]
```diff
@@ -23,13 +23,18 @@ class LSTMForecaster(BasePytorchForecaster):
...
+        >>> forecaster = LSTMForecaster.from_tsdataset(tsdata, **kwargs)
+        >>> forecaster.fit(tsdata, epochs=2, ...)
         >>> forecaster.to_local()  # if you set distributed=True
-        >>> test_pred = forecaster.predict(x_test)
+        >>> test_pred = forecaster.predict(x_test) # x_test also can be tsdata
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/forecaster/lstm_forecaster.py:37]: # x_test also can be tsdata can be deleted

File: python/chronos/test/bigdl/chronos/forecaster/test_lstm_forecaster.py [hunk 7]
```diff
@@ -60,6 +60,26 @@ def get_x_y(num_samples):
...
 
 
+def create_tsdataset(roll=True):
+    from bigdl.chronos.data import TSDataset
+    import pandas as pd
+    timeseries = pd.date_range(start='2020-01-01', freq='D', periods=1000)
+    df = pd.DataFrame(np.random.rand(1000, 2),
+                      columns=['value1', 'value2'],
+                      index=timeseries,
+                      dtype=np.float32)
```
Comment (TheaperDeng) [inline python/chronos/test/bigdl/chronos/forecaster/test_lstm_forecaster.py:70]: is there any error if not claiming dtype=np.float32, if yes, where (raise an issue)

File: python/chronos/src/bigdl/chronos/data/tsdataset.py [hunk 8]
```diff
@@ -49,8 +49,8 @@ def __init__(self, data, **schema):
 
         self.numpy_x = None
         self.numpy_y = None
-        self.roll_feature = None
-        self.roll_target = None
+        self.roll_feature = None  # Contains all feature_col column names
+        self.roll_target = None  # Contains all target_col column names
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/data/tsdataset.py:53]: contains target_col requested by `roll`/`to_torch_data_loader`

File: python/chronos/src/bigdl/chronos/forecaster/base_forecaster.py [hunk 9]
```diff
@@ -222,6 +223,14 @@ def fit(self, data, epochs=1, batch_size=32):
                | should be the same as past_seq_len and input_feature_num.
                | y's shape is (num_samples, horizon, target_dim), where horizon and target_dim
                | should be the same as future_seq_len and output_feature_num.
+               |
+               | 4. A bigdl.chronos.data.tsdataset.TSDataset instance:
+               | TSDataset provides some methods to prepare the dataset.
+               | If `TSDataset.roll` is called, the training speed will be faster,
+               | but but will consume more memory.
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/forecaster/base_forecaster.py:230]: one "but"

File: python/chronos/src/bigdl/chronos/forecaster/base_forecaster.py [hunk 10]
```diff
@@ -914,6 +989,58 @@ def quantize(self, calib_data=None,
...
+        """
+        Build a Forecaster Model.
+
+        :param tsdataset: A bigdl.chronos.data.tsdataset.TSDataset instance.
+        :param past_seq_len: Specify the history time steps (i.e. lookback).
+               Do not specify the 'past_seq_len' if your tsdataset has called
+               the 'TSDataset.roll' method.
+        :param future_seq_len: Specify the output time steps (i.e. horizon).
+               Do not specify the 'future_seq_len' if your tsdataset has called
+               the 'TSDataset.roll' method.
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/forecaster/base_forecaster.py:1003]: or `TSDataset.to_torch_data_loader`

File: python/chronos/src/bigdl/chronos/forecaster/base_forecaster.py [hunk 11]
```diff
@@ -914,6 +989,58 @@ def quantize(self, calib_data=None,
...
+        :param kwargs: Specify parameters of Forecaster,
+               e.g. loss and optimizer, etc.
+
+        :return: A Forecaster Model.
+        """
+        if isinstance(tsdataset, TSDataset):
+            output_feature_num = len(tsdataset.target_col)
+            if isinstance(tsdataset.roll_additional_feature, list):
+                tsdataset.feature_col += tsdataset.roll_additional_feature
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/forecaster/base_forecaster.py:1012]: I don't think you should change `tsdataset.feature_col`

File: python/chronos/src/bigdl/chronos/forecaster/base_forecaster.py [hunk 12]
```diff
@@ -222,6 +223,14 @@ def fit(self, data, epochs=1, batch_size=32):
...
                | y's shape is (num_samples, horizon, target_dim), where horizon and target_dim
                | should be the same as future_seq_len and output_feature_num.
+               |
+               | 4. A bigdl.chronos.data.tsdataset.TSDataset instance:
+               | TSDataset provides some methods to prepare the dataset.
+               | If `TSDataset.roll` is called, the training speed will be faster,
+               | but but will consume more memory.
+               | Otherwise, the training speed may be slower,
+               | but the memory usage will be reduced.
+               | Returns a DataLoader if `TSDataset.to_torch_data_loader` is called.
```
[the remark this answers is not in the document]
Comment (liangs6212) [reply python/chronos/src/bigdl/chronos/forecaster/base_forecaster.py:233]: I think that when the user does not understand the related operations of `TSDataset`, they can call `to_torch_data_loader` method and input the returned result into `fit` or other methods.

File: python/chronos/src/bigdl/chronos/forecaster/base_forecaster.py [hunk 13]
```diff
@@ -914,6 +990,49 @@ def quantize(self, calib_data=None,
...
+               the 'TSDataset.roll' method.
+        :param kwargs: Specify parameters of Forecaster,
+               e.g. loss and optimizer, etc.
+               More info, please refer to Forecaster.__init__ methods.
+
+        :return: A Forecaster Model.
+        """
+        if isinstance(tsdataset, TSDataset):
+            output_feature_num = len(tsdataset.target_col)
+            # TODO Support for gen_rolling_feature will be split into next pr
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/forecaster/base_forecaster.py:1013]: self.xxx is not None: report an error

File: python/chronos/src/bigdl/chronos/data/utils/roll_dataset.py [hunk 14]
```diff
@@ -93,6 +93,7 @@ def __init__(self, df, dt_col, freq, lookback, horizon, feature_col, target_col,
         self.lookback = lookback
         self.horizon = horizon
         self.target_num = len(target_col)
+        self.all_feature_num = len(feature_col + target_col)
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/data/utils/roll_dataset.py:96]: delete

File: python/chronos/src/bigdl/chronos/data/tsdataset.py [hunk 15]
```diff
@@ -49,8 +49,10 @@ def __init__(self, data, **schema):
 
         self.numpy_x = None
         self.numpy_y = None
-        self.roll_feature = None
-        self.roll_target = None
+        self.lookback = None
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/data/tsdataset.py:52]: add some comment such as
```python
# lookback stated by users if they called `roll`, `to_torch_data_loader`
```

File: python/chronos/src/bigdl/chronos/data/tsdataset.py [hunk 16]
```diff
@@ -778,6 +782,11 @@ def to_torch_data_loader(self,
                                         time_enc=time_enc,
                                         label_len=label_len,
                                         is_predict=is_predict)
+            horizon = torch_dataset.horizon
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/data/tsdataset.py:785]: I think we don't need this line? we have `horizon` in this method

File: python/chronos/src/bigdl/chronos/forecaster/base_forecaster.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (TheaperDeng) [reply python/chronos/src/bigdl/chronos/forecaster/base_forecaster.py:233]:
```python
'''
| 4. A bigdl.chronos.data.tsdataset.TSDataset instance:
| Forecaster will automatically process the TSDataset.
| By default, TSDataset will be transformed to a pytorch dataloader,
| which is memory-friendly while a little bit slower.
| Users may call `roll` on the TSDataset before calling `fit`
| Then the training speed will be faster but will consume more memory.
'''
```

File: python/chronos/src/bigdl/chronos/data/tsdataset.py [hunk 17]
```diff
@@ -49,8 +49,10 @@ def __init__(self, data, **schema):
 
         self.numpy_x = None
         self.numpy_y = None
-        self.roll_feature = None
-        self.roll_target = None
+        self.lookback = None
+        self.horizon = None
+        self.roll_feature = None  # contains feature_col requested by roll/to_torch_data_loader
+        self.roll_target = None  # contains target_col requested by roll/to_torch_data_loader
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/data/tsdataset.py:55]: please set `roll_feature` and `roll_target` in `to_torch_data_loader` as well

File: python/chronos/src/bigdl/chronos/forecaster/base_forecaster.py [hunk 18]
```diff
@@ -232,6 +241,13 @@ def fit(self, data, epochs=1, batch_size=32):
         :return: Evaluation results on data.
         """
         # input transform
+        if isinstance(data, TSDataset):
+            # if data.numpy_x is None, Need to call the roll method.
+            _rolled = data.numpy_x is None
+            data = data.to_torch_data_loader(batch_size=batch_size,
+                                             roll=_rolled,
+                                             lookback=self.data_config['past_seq_len'],
+                                             horizon=self.data_config['future_seq_len'])
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/forecaster/base_forecaster.py:250]:
```python
feature_col = data.roll_feature,
target_col = data.roll_target
```

File: python/chronos/src/bigdl/chronos/forecaster/base_forecaster.py [hunk 19]
```diff
@@ -914,6 +990,57 @@ def quantize(self, calib_data=None,
...
+
+            # calling roll or to_torch_data_loader
+            if tsdataset.horizon is not None:
+                past_seq_len = tsdataset.lookback
+                future_seq_len = tsdataset.horizon
+
+            # calling roll only
+            if tsdataset.numpy_x is not None:
+                output_feature_num = len(tsdataset.roll_target)
+                input_feature_num = output_feature_num + len(tsdataset.roll_feature)
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/forecaster/base_forecaster.py:1023]: we don't need line 1021-1023

**Review by TheaperDeng — COMMENTED** [review]
well check unittest later

**Review by TheaperDeng — APPROVED** [review]
LGTM, pls run jenkins

**Comment by liangs6212** [comment]
http://10.112.231.51:18889/view/BigDL-PR-Validation/job/BigDL-Chronos-PR-Validation/578/

**Comment by liangs6212** [comment]
http://10.112.231.51:18889/view/BigDL-PR-Validation/job/BigDL-Chronos-PR-Validation/588/

**Pull request merged**